File: src/rs/cache/CacheSystem.ts
import { CompressionHandler } from "../compression/CompressionHandler";
import { XteaKey } from "../crypto/Xtea";
import { CacheIndex, Dat2CacheIndex, DatCacheIndex } from "./CacheIndex";
import { CacheType } from "./CacheType";
import { Archive } from "./format/Archive";
import { ArchiveReference } from "./reference/ArchiveReference";
import { CacheStore } from "./store/CacheStore";

export class CacheSystem {
    static loadIndicesFromStore(
        cacheType: CacheType,
        store: CacheStore,
        indexIds: number[],
        compressionHandler: CompressionHandler,
    ): Map<number, CacheIndex> {
        const indices: Map<number, CacheIndex> = new Map();

        for (const id of indexIds) {
            const index =
                cacheType === CacheType.Dat
                    ? DatCacheIndex.fromStore(id, store, compressionHandler)
                    : Dat2CacheIndex.fromDat2Store(id, store, compressionHandler);
            indices.set(id, index);
        }

        return indices;
    }

    static fromStore(
        cacheType: CacheType.Dat | CacheType.Dat2,
        store: CacheStore,
        indexIds: number[],
        compressionHandler: CompressionHandler,
    ): CacheSystem {
        const indices = CacheSystem.loadIndicesFromStore(
            cacheType,
            store,
            indexIds,
            compressionHandler,
        );
        return new CacheSystem(indices, compressionHandler);
    }

    constructor(
        readonly indices: ReadonlyMap<number, CacheIndex>,
        readonly compressionHandler: CompressionHandler,
    ) {}

    indexExists(indexId: number): boolean {
        return this.indices.has(indexId);
    }

    tryGetIndex(indexId: number): CacheIndex | undefined {
        return this.indices.get(indexId);
    }

    getIndex(indexId: number): CacheIndex {
        const index = this.indices.get(indexId);
        if (!index) {
            throw new Error("Index not found: " + indexId);
        }
        return index;
    }

    readArchiveBytes(indexId: number, archiveId: number): Uint8Array {
        return this.getIndex(indexId).readArchiveBytes(archiveId);
    }

    readContainerPayload(indexId: number, archiveId: number, key: XteaKey | null): Uint8Array {
        return this.getIndex(indexId).readContainerPayload(archiveId, key);
    }

    tryReadContainerPayload(
        indexId: number,
        archiveId: number,
        key: XteaKey | null,
    ): Uint8Array | undefined {
        return this.indices.get(indexId)?.tryReadContainerPayload(archiveId, key);
    }

    getArchiveKey(indexId: number, archiveId: number, key: XteaKey | null): Archive {
        return this.getIndex(indexId).getArchiveKey(archiveId, key);
    }

    getArchive(indexId: number, archiveId: number): Archive {
        return this.getIndex(indexId).getArchive(archiveId);
    }

    tryGetArchive(indexId: number, archiveId: number): Archive | undefined {
        return this.indices.get(indexId)?.tryGetArchive(archiveId);
    }

    /**
     * Dat2-only: returns the archive reference/metadata for splitting archive payload into files
     * (and optional integrity/name fields). Returns null if not available.
     */
    getArchiveReference(indexId: number, archiveId: number): ArchiveReference | null {
        return this.getIndex(indexId).getArchiveReference(archiveId);
    }

    /**
     * Dat2-only: returns the subset of archive metadata required to split payload bytes into files.
     * Returns null if not available.
     */
    getArchiveMeta(
        indexId: number,
        archiveId: number,
    ): {
        id: number;
        lastFileId: number;
        fileCount: number;
        fileIds: Int32Array;
        fileNameHashes: Int32Array;
    } | null {
        const ref = this.getArchiveReference(indexId, archiveId);
        if (!ref) {
            return null;
        }
        return {
            id: ref.id,
            lastFileId: ref.lastFileId,
            fileCount: ref.fileCount,
            fileIds: ref.fileIds,
            fileNameHashes: ref.fileNameHashes,
        };
    }
}

File: cpp/rs/cache/CacheSystem.cpp
#include "CacheSystem.hpp"

#include <cstddef>

#include "../cache/ArchiveMeta.hpp"
#include "../compression/CompressionHandler.hpp"
#include "../core/Allocator.hpp"
#include "../core/Move.hpp"
#include "../core/Result.hpp"
#include "../core/Span.hpp"
#include "../core/Status.hpp"
#include "../core/Vec.hpp"
#include "../crypto/Xtea.hpp"
#include "../types.hpp"
#include "CacheIndex.hpp"
#include "CacheType.hpp"
#include "format/Archive.hpp"
#include "store/CacheStore.hpp"

namespace rs {

Result<CacheSystem> CacheSystem::fromStore(
    CacheType cacheType,
    const CacheStore& store,
    Span<const i32> indexIds,
    const CompressionHandler& compressionHandler,
    Allocator& alloc) noexcept {
    Vec<CacheIndex> indices(alloc);

    auto r = indices.reserve(indexIds.size());
    if (!r.isOk()) {
        return Result<CacheSystem>::err(r.status());
    }

    for (std::size_t i = 0; i < indexIds.size(); i++) {
        const i32 id = indexIds[i];
        auto indexRes = CacheIndex::fromStore(cacheType, id, store, compressionHandler, alloc);
        if (!indexRes.isOk()) {
            return Result<CacheSystem>::err(indexRes.status());
        }
        (void)indices.pushBack(rs::move(indexRes.value()));
    }

    return Result<CacheSystem>::ok(CacheSystem(rs::move(indices), compressionHandler));
}

bool CacheSystem::indexExists(i32 indexId) const noexcept {
    for (std::size_t i = 0; i < indices_.size(); i++) {
        if (indices_[i].id() == indexId) {
            return true;
        }
    }
    return false;
}

Status CacheSystem::getIndex(i32 indexId, const CacheIndex** out) const noexcept {
    if (!out) {
        return Status::InvalidArgument;
    }
    for (std::size_t i = 0; i < indices_.size(); i++) {
        if (indices_[i].id() == indexId) {
            *out = &indices_[i];
            return Status::Ok;
        }
    }
    *out = nullptr;
    return Status::NotFound;
}

Status CacheSystem::readArchiveBytes(i32 indexId, i32 archiveId, Vec<u8>* out, Allocator& alloc) const noexcept {
    const CacheIndex* index = nullptr;
    const Status s = getIndex(indexId, &index);
    if (!ok(s)) {
        return s;
    }
    return index->readArchiveBytes(archiveId, out, alloc);
}

Status CacheSystem::readContainerPayload(i32 indexId, i32 archiveId, const XteaKey* key, Vec<u8>* out, Allocator& alloc) const noexcept {
    const CacheIndex* index = nullptr;
    const Status s = getIndex(indexId, &index);
    if (!ok(s)) {
        return s;
    }
    return index->readContainerPayload(archiveId, key, out, alloc);
}

Status CacheSystem::getArchiveMeta(i32 indexId, i32 archiveId, ArchiveMeta* out) const noexcept {
    const CacheIndex* index = nullptr;
    const Status s = getIndex(indexId, &index);
    if (!ok(s)) {
        return s;
    }
    return index->getArchiveMeta(archiveId, out);
}

Result<Archive> CacheSystem::getArchiveKey(i32 indexId, i32 archiveId, const XteaKey* key, Allocator& alloc) const noexcept {
    const CacheIndex* index = nullptr;
    const Status s = getIndex(indexId, &index);
    if (!ok(s)) {
        return Result<Archive>::err(s);
    }
    return index->getArchiveKey(archiveId, key, alloc);
}

} // namespace rs
